### PR TITLE
Fixes #1592: making report::expire faster and without errors

### PR DESCRIPTION
**Pull request by orrabin**

### CODE REVIEW

**Comment by orrabin** [comment]
I'd like to test it on real reports but for now on 5000 generated reports I used:
`puts Benchmark.measure {Report.expire()}`
and got these results:

before change:
2014-08-07 10:20:09 +0300: Expired 5000 Reports
  0.470000   0.000000   0.470000 (  1.963282)

after change:
2014-08-07 09:52:39 +0300: Expired 5000 Reports
  0.020000   0.010000   0.030000 (  0.483464)

**Comment by domcleal** [comment]
Something else to measure will be the memory usage, as that's been a common complaint about reports:expire and it looks like the original code was attempting to keep it down (successfully or not, I'm unsure).  `/bin/time` can measure max resident size, which may be useful.

**Comment by orrabin** [comment]
@domcleal I used `/bin/time` as you suggested and the max resident size seems to be better after my changes. Are there more tests I can run? 

 The results:
before change:
14.95user 0.90system 0:17.39elapsed 91%CPU (0avgtext+0avgdata 751664maxresident)k
0inputs+1328outputs (0major+118227minor)pagefaults 0swaps

after change:
14.14user 0.91system 0:15.53elapsed 96%CPU (0avgtext+0avgdata 584480maxresident)k
0inputs+64outputs (0major+105726minor)pagefaults 0swaps

File: app/models/report.rb [hunk 1]
```diff
@@ -117,44 +117,15 @@ def <=>(other)
...
-    Log.where(:report_id => orphaned_logs).delete_all unless orphaned_logs.empty?
-
-    all_messages = Message.pluck(:id)
-    orphaned_messages = all_messages - used_messages
-    Message.where(:id => orphaned_messages).delete_all unless orphaned_messages.empty?
-
-    all_sources = Source.pluck(:id)
-    orphaned_sources = all_sources - used_sources
-    Source.where(:id => orphaned_sources).delete_all unless orphaned_sources.empty?
+    count = 0
```
Comment (dLobatog) [inline app/models/report.rb]: Why initializing it to 0?

`count = Report.where(cond).delete_all` should do.

File: app/models/report.rb [hunk 2]
```diff
@@ -117,44 +117,15 @@ def <=>(other)
...
+    count += Report.where(cond).delete_all
 
     logger.info Time.now.to_s + ": Expired #{count} Reports"
     return count
```
Comment (dLobatog) [inline app/models/report.rb]: No need for the return keyword here, just use `count` and the convention of the last statement, this is code is a little legacy.

**Comment by dLobatog** [comment]
:+1: except for the code nitpicks, even if it was a tad slower I'd favor this as it's considerably simpler to understand and maintain.

**Comment by orrabin** [comment]
@eLobato sorry, those where leftovers from the previous code.
Fixed the nitpicks :)

**Comment by dLobatog** [comment]
I'm testing memory usage for 10k reports before merging.

**Comment by orrabin** [comment]
@eLobato thanks!

**Comment by dLobatog** [comment]
Peak mem usage seems to be even lower here than in the old version -

Merged as https://github.com/theforeman/foreman/commit/4215defb81fc2b469669592ac1094d7c70950192 , thanks @orrabin !

File: app/models/report.rb [hunk 3]
```diff
@@ -117,47 +117,16 @@ def <=>(other)
...
-    all_messages = Message.pluck(:id)
-    orphaned_messages = all_messages - used_messages
-    Message.where(:id => orphaned_messages).delete_all unless orphaned_messages.empty?
+    cond = "reports.created_at < \'#{(Time.now.utc - timerange).to_formatted_s(:db)}\'"
+    cond += " and reports.status = #{status}" unless status.nil?
 
-    all_sources = Source.pluck(:id)
-    orphaned_sources = all_sources - used_sources
-    Source.where(:id => orphaned_sources).delete_all unless orphaned_sources.empty?
+    Message.joins(:logs).joins(:reports).where(cond)
```
Comment (ohadlevy) [inline app/models/report.rb]: unused query

File: app/models/report.rb [hunk 4]
```diff
@@ -117,47 +117,16 @@ def <=>(other)
...
-    all_reports.uniq! ; used_messages.uniq! ; used_sources.uniq!
-
-    # reports which have logs entries
-    used_reports = Report.where(:id => all_reports).pluck(:id)
-
-    orphaned_logs = all_reports - used_reports
-    Log.where(:report_id => orphaned_logs).delete_all unless orphaned_logs.empty?
-
-    all_messages = Message.pluck(:id)
-    orphaned_messages = all_messages - used_messages
```
Comment (ohadlevy) [inline app/models/report.rb:152]: this patch lacks the orphaned messages and sources support :)

**Comment by orrabin** [comment]
@eLobato these are the test results for 10k

before change:
2014-08-09 19:54:08 +0300: Expired 10000 Reports
  1.050000   0.020000   1.070000 (  4.305314)
17.27user 1.19system 0:20.93elapsed 88%CPU (0avgtext+0avgdata 661616maxresident)k
0inputs+1456outputs (0major+109556minor)pagefaults 0swaps

after change:
2014-08-09 20:39:47 +0300: Expired 10000 Reports
  0.080000   0.000000   0.080000 ( 12.673412)
13.98user 0.82system 0:22.16elapsed 66%CPU (0avgtext+0avgdata 591392maxresident)k
80inputs+72outputs (1major+106172minor)pagefaults 0swaps

**Comment by ohadlevy** [comment]
[test]

**Comment by orrabin** [comment]
[test]

**Comment by orrabin** [comment]
@eLobato can you please recheck this?

**Comment by ohadlevy** [comment]
merged as ab866a3 thanks @orrabin

**Pull request closed without merging**